src/plugins/sync_to_async.ts

```diff
@@ -0,0 +1,380 @@
+        // v1 api, then we will create a new binding in the scope, and store a
+        // reference to it in the cache, i.e. `{ ticket: "_ticket2"  }`.
+        id = opCache.scope.generateUidIdentifier(name);
+        // This hoists the new variable to the method scope
+        opCache.scope.push({
+          id,
+          kind: "const",
+          init: buildWrapZafClientExpression(name)
+        });
+        opCache[name] = id;
+        toAsync = true;
+      }
+      // If we previously created a binding and hoisted, then we need to
+      // use that binding to replace an expression.
+      if (id) {
+        // If names has any values, then we need to replace a wrapper expression
+        // direct property access, i.e. instead of `this.ticket().requester().email()`
+        // we would replace with `_ticket.requester.email` (where `_ticket` was a new
+        // hoisted variable we created).
+        if (names.length) {
+          // Just concat the binding name with property names
+          exp.replaceWithSourceString(`${id.name}.${names.join(".")}`);
+          // If this expression is itself an assignment to a binding, we need
+          // to continue processing further references, i.e.
+          // `var ticket = this.ticket(); var requester = ticket.requester();`
+          if (exp.parentPath.isVariableDeclarator()) {
+            const binding = opCache.scope.bindings[exp.parent.id.name];
+            replaceReferencesForBinding(binding);
+          }
+        } else {
+          // If we get to here, we're just using the value of the binding
+          // directly, i.e. `return this.ticket()`.
+          exp.replaceWith(id);
+        }
+        toAsync = true;
+      }
+    }
+    // If we didn't set `toAsync`, then we haven't found/replaced any expressions.
+    if (!toAsync) return;
+    // If `asyncMethods` exists and this app method isn't in the queue to be processed
+    // in the second pass, add it, since it will become async.
+    if (asyncMethods && !asyncMethods.includes(opName)) {
+      asyncMethods.push(opName);
+    }
+    // Set the FunctionExpression node to be async.
+    op.node.value.async = true;
+    // Keep a reference to the method being async in metadata for lookup later
+    opCache.async = true;
+  }
+};
+
+interface IManifest {
+  [key: string]: any; // number | object | string | boolean
+  location?: string | string[];
+  noTemplate?: string | string[];
+}
+
+/**
+ * Checks whether the given location is specified in the manifest.
+ * A location could appear in either the `location`, or the `noTemplate`.
+ * 
+ * @param {IManifest} manifest 
+ * @param {string} name 
+ * @returns {boolean} 
+ */
+function hasLocation(manifest: IManifest, name: string): boolean {
+  // Store the intersection of all the possible locations in a
+  // new transient property on the manifest object.
+  if (!manifest.allLocations) {
+    let { location, noTemplate } = manifest;
```

Comment:
Why do we care about `noTemplate`? Even if a location is `noTemplate` it still needs to be listed under `location`, no?

Reply:
I'll simplify this function as per your suggestion